src/sentry/middleware/integrations/classifications.py

```diff
@@ -0,0 +1,114 @@
+    PluginRequestParser,
+    SlackRequestParser,
+    VstsRequestParser,
+)
+
+if TYPE_CHECKING:
+    from sentry.middleware.integrations.integration_control import ResponseHandler
+
+    from .parsers.base import BaseRequestParser
+
+
+class BaseClassification(abc.ABC):
+    def __init__(self, response_handler: ResponseHandler) -> None:
+        self.response_handler = response_handler
+
+    def should_operate(self, request: HttpRequest) -> bool:
+        """
+        Determines whether the classification should act on request.
+        Middleware will return self.get_response() if this function returns True.
+        """
+        raise NotImplementedError
+
+    def get_response(self, request: HttpRequest) -> HttpResponse:
+        """Parse the request and return a response."""
+        raise NotImplementedError
+
+
+class PluginClassification(BaseClassification):
+    def should_operate(self, request: HttpRequest) -> bool:
+        rp = PluginRequestParser(request=request, response_handler=self.response_handler)
+        return rp.should_operate()
+
+    def get_response(self, request: HttpRequest) -> HttpResponse:
+        rp = PluginRequestParser(request=request, response_handler=self.response_handler)
+        return rp.get_response()
+
+
+class IntegrationClassification(BaseClassification):
+    integration_prefix: str = "/extensions/"
+    """Prefix for all integration requests. See `src/sentry/web/urls.py`"""
+    setup_suffix: str = "/setup/"
+    """Suffix for PipelineAdvancerView on installation. See `src/sentry/web/urls.py`"""
+    logger = logging.getLogger(f"{__name__}.integration")
+    active_parsers: List[Type[BaseRequestParser]] = [
+        BitbucketRequestParser,
+        BitbucketServerRequestParser,
+        GithubEnterpriseRequestParser,
+        GithubRequestParser,
+        GitlabRequestParser,
+        JiraRequestParser,
+        JiraServerRequestParser,
+        MsTeamsRequestParser,
+        SlackRequestParser,
+        VstsRequestParser,
+    ]
+    integration_parsers: Mapping[str, Type[BaseRequestParser]] = {
+        cast(str, parser.provider): parser for parser in active_parsers
+    }
+
+    def _identify_provider(self, request: HttpRequest) -> str | None:
+        """
+        Parses the provider out of the request path
+            e.g. `/extensions/slack/commands/` -> `slack`
+        """
+        integration_prefix_regex = re.escape(self.integration_prefix)
+        provider_regex = rf"^{integration_prefix_regex}(\w+)"
+        result = re.search(provider_regex, request.path)
+        if not result:
+            self.logger.error("invalid_provider", extra={"path": request.path})
+            return None
+        return result.group(1)
+
+    def should_operate(self, request: HttpRequest) -> bool:
+        is_integration = request.path.startswith(self.integration_prefix)
+        is_not_setup = not request.path.endswith(self.setup_suffix)
+        return is_integration and is_not_setup
```

Comment:
it's marginally faster to write this as a single statement such that short circuit evaluation could skip the second -- though startswith is pretty fast so prob ably not worth it